corehq/apps/accounting/models.py

```diff
@@ -2649,18 +2649,34 @@ def get_credits_for_line_item(cls, line_item):
             line_item.feature_rate.feature.feature_type
             if line_item.feature_rate is not None else None
         )
-        return itertools.chain(
-            cls.get_credits_by_subscription_and_features(
+
+        for credit_line in cls.get_credits_by_subscription_and_features(
+            line_item.invoice.subscription,
+            product_type=product_type,
+            feature_type=feature_type,
+        ):
+            yield credit_line
+
+        if product_type is not None:
+            for credit_line in cls.get_credits_by_subscription_and_features(
                 line_item.invoice.subscription,
-                product_type=product_type,
-                feature_type=feature_type,
-            ),
-            cls.get_credits_for_account(
+                product_type=SoftwareProductType.ANY,
+            ):
+                yield credit_line
+
+        for credit_line in cls.get_credits_for_account(
+            line_item.invoice.subscription.account,
+            product_type=product_type,
+            feature_type=feature_type,
+        ):
+            yield credit_line
+
+        if product_type is not None:
+            for credit_line in cls.get_credits_for_account(
                 line_item.invoice.subscription.account,
-                product_type=product_type,
-                feature_type=feature_type,
-            )
-        )
+                product_type=SoftwareProductType.ANY,
+            ):
+                yield credit_line
```

Comment:
There's a lot going on in this function, and it would be nice to refactor it into separate functions for product and feature credit lines.  But I'm going to punt this to a separate PR.